File: support-app/src/main/java/com/support_app/service/DefectService.java
package com.support_app.service;

import com.support_app.dto.CreateDefectDto;
import com.support_app.dto.UpdateDefectDto;
import com.support_app.exception.DefectNotFoundException;
import com.support_app.mapper.DefectMapper;
import com.support_app.model.Defect;
import com.support_app.repository.DefectRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class DefectService {

    @Autowired
    private DefectRepository defectRepository;

    @Autowired
    private DefectMapper defectMapper;

    /**
     * Retrieves all defects from the system.
     *
     * @return A list of all Defect entities.
     */
    public List<Defect> getAllDefects() {
        return defectRepository.findAll();
    }

    /**
     * Adds a new defect to the system.
     *
     * @param dto The DTO containing the details of the defect to be created.
     * @return The created Defect entity.
     */
    public Defect addDefect(CreateDefectDto dto) {
        Defect defect = defectMapper.fromCreateDto(dto);
        return defectRepository.save(defect);
    }

    /**
     * Updates an existing defect in the system.
     *
     * @param dto The DTO containing the updated details of the defect.
     * @return The updated Defect entity.
     * @throws DefectNotFoundException If a defect with the given ID is not found.
     */
    public Defect updateDefect(UpdateDefectDto dto) {
        Defect defect = defectRepository.findById(dto.getId())
                .orElseThrow(() -> new DefectNotFoundException(dto.getId()));

        defectMapper.updateDefectFromDto(dto, defect);
        return defectRepository.save(defect);
    }

    /**
     * Deletes a defect from the system.
     *
     * @param defectId The ID of the defect to delete.
     * @throws DefectNotFoundException If a defect with the given ID is not found.
     */
    public void deleteDefect(Long defectId) {
        Defect defect = defectRepository.findById(defectId)
                .orElseThrow(() -> new DefectNotFoundException(defectId));
        defectRepository.delete(defect);
    }
}

File: support-frontend/src/app/core/services/defect.service.ts
// src/app/core/services/defect.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import {Defect} from "../models/defect.model";
import {CreateDefectDto} from "../dtos/create-defect-dto.dto";
import {UpdateDefectDto} from "../dtos/update-defect-dto.dto";
import {environment} from "../../../environments/environment";

@Injectable({
  providedIn: 'root'
})
export class DefectService {

  private apiUrl = `${environment.apiUrl}/defects`;

  constructor(private http: HttpClient) { }

  getAllDefects(): Observable<Defect[]> {
    return this.http.get<Defect[]>(`${this.apiUrl}/all`);
  }

  addDefect(createDefectDto: CreateDefectDto): Observable<Defect> {
    return this.http.post<Defect>(`${this.apiUrl}/add`, createDefectDto);
  }

  updateDefect(updateDefectDto: UpdateDefectDto): Observable<Defect> {
    return this.http.put<Defect>(`${this.apiUrl}/update`, updateDefectDto);
  }

  deleteDefect(defectId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/delete/${defectId}`);
  }
}
